korge/src/commonMain/kotlin/com/soywiz/korge/view/Sprite.kt

```diff
@@ -39,68 +39,68 @@ inline fun Container.sprite(
  * This will be exchanged when starting a [SpriteAnimation] with one of the available play functions
  */
 open class Sprite(
-    bitmap : Bitmap,
+    bitmap: Bitmap,
     anchorX: Double = 0.0,
     anchorY: Double = anchorX,
     hitShape: VectorPath? = null,
     smoothing: Boolean = true) : Image(bitmap, anchorX, anchorY, hitShape, smoothing) {
     constructor(
-        bmpSlice : BmpSlice,
+        bmpSlice: BmpSlice,
         anchorX: Double = 0.0,
         anchorY: Double = anchorX,
         hitShape: VectorPath? = null,
         smoothing: Boolean = true) : this(bmpSlice.bmp, anchorX, anchorY, hitShape, smoothing)
+
     constructor(
-        initialAnimation : SpriteAnimation,
+        initialAnimation: SpriteAnimation,
         anchorX: Double = 0.0,
         anchorY: Double = anchorX,
         hitShape: VectorPath? = null,
-        smoothing: Boolean = true) : this(initialAnimation.firstSprite, anchorX, anchorY, hitShape, smoothing){
+        smoothing: Boolean = true) : this(initialAnimation.firstSprite, anchorX, anchorY, hitShape, smoothing) {
         currentAnimation = initialAnimation
         setFrame(0)
     }
 
     private var animationRequested = false
     private var animationNumberOfFramesRequested = 0
         set(value) {
-            if (value == 0)
-                triggerEvent(_onAnimationCompleted)
+            if (value == 0) {
+                stopAnimation()
+                when (animationType) {
+                    AnimationType.STANDARD -> triggerEvent(onAnimationCompleted)
+                    else -> triggerEvent(onAnimationStopped)
+                }
+            }
             field = value
         }
+    private var animationType = AnimationType.STANDARD
+    val onAnimationCompleted = Signal<SpriteAnimation>()
+    val onAnimationStopped = Signal<SpriteAnimation>()
+    val onAnimationStarted = Signal<SpriteAnimation>()
 
-    private var _onAnimationCompleted: Signal<SpriteAnimation>? = null
-    private var _onAnimationStopped: Signal<SpriteAnimation>? = null
-    private var _onAnimationStarted: Signal<SpriteAnimation>? = null
-
-    val onAnimationCompleted: Signal<SpriteAnimation>
-        get() {
-            if (_onAnimationCompleted == null) _onAnimationCompleted = Signal()
-            return _onAnimationCompleted!!
-        }
-    val onAnimationStopped: Signal<SpriteAnimation>
-        get() {
-            if (_onAnimationStopped == null) _onAnimationStopped = Signal()
-            return _onAnimationStopped!!
-        }
-
-    val onAnimationStarted: Signal<SpriteAnimation>
-        get() {
-            if (_onAnimationStarted == null) _onAnimationStarted = Signal()
-            return _onAnimationStarted!!
-        }
-
```

Comment:
This optimization was lost: https://github.com/korlibs/korge/pull/158/commits/5aab8def22baa39b4b51dc824aaa2c06419645d0

Some Sprite won't have subscribers, and you are creating several objects and doing iterations when dispatching them even with no subscribers. The 10K Sprites sample benefitied from this optimization

Reply:
I will reintroduce the changes. Have they been Public already ?

Reply:
They are on master already but not published a version yet. You probably forked the repo before that was merged

Reply:
Seems so. Well. Will be fixed :)